brave/brave-core
# [iOS] Search Suggestions Enhancement

Resolves https://github.com/brave/brave-browser/issues/38137

Security Ticket: https://github.com/brave/reviews/issues/1567 to evaluate the spec
https://github.com/brave/reviews/issues/1672 to evaluate the implementation iOS

The main focus:

This PR is aiming to enhance the user experience especially for the first time users for search suggestions. Some users are confused about enabling search suggestions and considering Brave doesn't have search suggestions because they can easily miss prompt for enabling it or forget about disabling it and don't know where to enable inside setting screens later.

To improve this, users from Brave Search Default regions have their search suggestions enabled by default. (They can disable from settings etc, that behaviour is not changed)

The change can be seen at `InitialSearchEngine` and `SearchEngine` classes along with the tests for it inside `InitialSearchEngineTests`.

To enabled this behaviour the additional changes were required

- Do not offer suggestions if you have pasted from clipboard and isSuspicious query method . After 1 keystroke the suggestions should appear. And test the pasted query with this method.

- `Suspicious Query` check is converted into swift from https://github.com/brave/search/blob/main/services/mixer/serpkit/src/lib/utils/hw.ts#L376

and the tests from https://github.com/brave/search/blob/a68168a840980b219c4ed9ea4cc45f3550ea6296/services/mixer/serpkit/tests/lib/utils/hw.test.ts#L95 are also added to prove the converted function acts as desired.

This conversion also included the conversion regex phrases to be used in swift with rules like 

-Use a "raw" string literal to avoid double escaping backslashes that form regex escapes (you have \1 backreference and \. regex escape in the pattern), so use let format = #"..."#

-Escape square brackets inside character classes in a regex pattern because square brackets "[" and "]" are both special (they are used to create character class intersections and unions).

Also some quality of life changes are made to SearchEngines class.



## Submitter Checklist:

  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`

## Reviewer Checklist:


## After-merge Checklist:

  changes has landed on

## Test Plan:

- Change Region to one of the Brave Search enabled default countries like Canada or USA
- Fresh Install Brave
- Click on URL Bar and check the search suggestion enable-disable table section is not there (it should be enabled by default)
- Using keyboard type `test` and check search suggestions are displayed and delete the entry
- Copy paste "test query" and check search suggestions are displayed (this is one of the safe queries)
- Copy paste "test query  \t \t \t \t \t \t \t \t \t \t \t \t \t \t \t \t \t \t \t \t \t \t \t \t \t \t \t \t \t \t \t \t \t \t \t \t \t \t \t \t \t" and check search queries are not added( also this is one of the suspicious queries in the list)

## Screenshots-Videos:


https://github.com/brave/brave-core/assets/6643505/e9a69a89-4518-482e-af84-bef6b9484ca5
